### PR TITLE
Add `.modal-dialog-centered` for optional vertically centered modal

### LINKED ISSUE #23638: v4 - Modal align vertical center
I need to leave my modal aligned to the center of the screen vertically.

I tried with Flex and with Translate, however the 2 solutions cause bizarre behavior in the modal.

Is there any native form of bootstrap v4 doing this or some workround?

Several workround's found have problem when content is large.

### PR DESCRIPTION
**Pull request by mdo**
Fixes #23638

### CODE REVIEW

**Comment by andresgalante** [comment]
@mdo I've added a commit to remove margin on responsive states too to remove vertical scrolls and added the centered class to the example on the docs.
@mdo @XhmikosR Can you please remove my changes?

**Comment by XhmikosR** [comment]
@andresgalante: what do you mean remove? Why don't you revert your patch?

**Comment by andresgalante** [comment]
Sorry :) my patch is correct, I wrote "review" and autocorrect changed it

File: docs/4.0/components/modal.md [hunk 1]
```diff
@@ -208,6 +208,64 @@ When modals become too long for the user's viewport or device, they scroll indep
 </div>
 {% endhighlight %}
 
+### Vertically centered
+
+Add `.modal-dialog-centered` to `.modal-dialog` to vertically center the modal.
+
+<div id="exampleModalCenter" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="exampleModalCenter" aria-hidden="true">
```
Comment (tmorehouse) [inline docs/4.0/components/modal.md]: should `aria-labeledby` be set to `"exampleModalCenterTitle"`?

**Comment by tmorehouse** [comment]
One other thing that might be worth noting is that `.modal-dialog-centered` is only suitable for short modals (i.e. one or two liner content).  Tall modals will have their header cut off (unable to click the top close button, and no visible `.modal-title`), although the footer is accessible via scrolling.

**Comment by andresgalante** [comment]
Hey @tmorehouse you are right, once the modal is larger than the viewport it cut off the header.

I tried to find a way out of this, removing the 100% height on`modal-dialog-centered` does it but of course it doesn't center it.

To be honest I don't think it's an option to document it as suitable for short texts, a component should be robust enough to handle anything. You solution is smart, I like it! I want to try find a way to keep our scroll. So far I failed with every attempt I went for 😄  But I'll try a bit more. 

Thanks a lot for reviewing it and finding the gaps!

**Comment by tmorehouse** [comment]
Yeah, I tried trying to get that header back into the viewport, but failed miserably as well trying to keep the right hand scrollbar at the same time.  My brain is not flex-y enough to figure it out (not without out adding additional wrapper elements). 

But the scrollable modal-body option might be a "simple" (air quotes) solution (for now) without requiring extra markup. And the existing modal JS code appears to work with it well. (it just doesn't add a scroll bar on the right of the viewport)

**Comment by XhmikosR** [comment]
I wonder, can't we use the old `translateY` trick?

**Comment by tmorehouse** [comment]
Would `translateY` interfere with the fade animation?

**Comment by XhmikosR** [comment]
I haven't had a close look into this, I was just thinking out loud, in case we missed something obvious.

**Comment by sesemaya** [comment]
@andresgalante 

Wouldn't `.modal-dialog-centered { min-height: 100%; }` instead of `height: 100%;` fix the header cut off problem while still keep smaller modals vertically centred?

**Comment by tmorehouse** [comment]
Another option might be javascript.  Just before modal opens, set modal-content height to 0, then measure the scrollHeight, and if less than the viewport height use the centered CSS, and if taller than the viewport, revert back to non-vertically centered (and be sure not to forget to reset height to auto before starting the fade in transition.

**Pull request merged**